pkg/envoy_access_log_server/service_v3.go

```diff
@@ -0,0 +1,225 @@
+		}
+	}
+
+	result["requestBodyBytes"] = strconv.FormatUint(request.RequestBodyBytes, 10)
+	result["requestHeadersBytes"] = strconv.FormatUint(request.RequestHeadersBytes, 10)
+	result["requestId"] = request.RequestId
+	result["requestMethod"] = request.RequestMethod.String()
+	result["scheme"] = request.Scheme
+	result["userAgent"] = request.UserAgent
+
+	return result
+}
+
+type envoyV3AccessLogEntryHttpResponseProperties envoy_data_accesslog_v3.HTTPResponseProperties
+
+func (response envoyV3AccessLogEntryHttpResponseProperties) StringMap() stringmap.StringMap {
+	result := stringmap.StringMap{}
+
+	result["responseBodyBytes"] = strconv.FormatUint(response.ResponseBodyBytes, 10)
+	result["responseCodeDetails"] = response.ResponseCodeDetails
+	result["responseCode"] = strconv.FormatUint(uint64(response.ResponseCode.GetValue()), 10)
+	result["responseHeadersBytes"] = strconv.FormatUint(response.ResponseHeadersBytes, 10)
+	// response headers are encoded with `sent_http_` prefix (to keep the scheme used by Nginx)
+	if response.ResponseHeaders != nil {
+		for header := range response.ResponseHeaders {
+			result["sent_http_"+header] = response.ResponseHeaders[header]
+		}
+	}
+	if response.ResponseTrailers != nil {
+		for trailer := range response.ResponseTrailers {
+			result["sent_trailer_"+trailer] = response.ResponseTrailers[trailer]
+		}
+	}
+	return result
+}
+
+type envoyV3HttpAccessLogEntry envoy_data_accesslog_v3.HTTPAccessLogEntry
+
+func (l envoyV3HttpAccessLogEntry) StringMap() stringmap.StringMap {
+	m := envoyV3AccessLogEntryCommonProperties(*l.CommonProperties).StringMap()
+	m["protocolVersion"] = l.ProtocolVersion.String()
+	m = m.Merge(envoyV3AccessLogEntryHttpRequestProperties(*l.Request).StringMap())
+	m = m.Merge(envoyV3AccessLogEntryHttpResponseProperties(*l.Response).StringMap())
+	return m
+}
+
+type envoyV3TcpAccessLogEntry envoy_data_accesslog_v3.TCPAccessLogEntry
+
+func (l envoyV3TcpAccessLogEntry) StringMap() stringmap.StringMap {
+	m := envoyV3AccessLogEntryCommonProperties(*l.CommonProperties).StringMap()
+	m["receivedBytes"] = strconv.FormatUint(l.ConnectionProperties.ReceivedBytes, 10)
+	m["sentBytes"] = strconv.FormatUint(l.ConnectionProperties.SentBytes, 10)
+	return m
+}
+
+func (service_v3 *AccessLogServiceV3) exportLogEntriesV3(msg *envoy_service_accesslog_v3.StreamAccessLogsMessage) []stringmap.StringMap {
+	var result []stringmap.StringMap
+
+	if logs := msg.GetHttpLogs(); logs != nil {
+		result = make([]stringmap.StringMap, len(logs.LogEntry))
+		for i, l := range logs.LogEntry {
+			result[i] = envoyV3HttpAccessLogEntry(*l).StringMap()
+			logEntriesTotal.WithLabelValues("HTTP", "v3").Inc()
+		}
+	} else if logs := msg.GetTcpLogs(); logs != nil {
+		result = make([]stringmap.StringMap, len(logs.LogEntry))
+		for i, l := range logs.LogEntry {
+			result[i] = envoyV3TcpAccessLogEntry(*l).StringMap()
+			logEntriesTotal.WithLabelValues("TCP", "v3").Inc()
+		}
```

Comment:
These two blocks of code are quite a duplication, maybe we can leave the iteration for both and use a type switch for each message? But I do get the optimization done here by knowing that the msg always contains entries of the same type :thinking:

Reply:
https://github.com/seznam/slo-exporter/pull/37#discussion_r557702086